File: PalindromePartitioningII.cpp
class Solution {
public:
    int minCut(string s) {
        int n = s.size();
        bool p[n][n] = {0};
        int cut[n + 1];
        for (int i = 0; i <= n; ++i)
            cut[i] = n - 1 - i;
        for (int i = n - 1; i >= 0; --i) {
            for (int j = i; j < n; ++j)
                if (s[i] == s[j] && (j - i < 2 || p[i + 1][j - 1])) {
                    p[i][j] = true;
                    cut[i] = min(cut[i], cut[j + 1] + 1);
                }
        }
        
        return cut[0];
    }
};

File: lulu/palindromePartitioningII.py
class Solution(object):
    def minCut(self, s):
        memo = [float('inf')] * len(s)
        memo[-1] = 0
        for start in range(len(s) - 2, -1, -1):
            if self.isPalindrome(start, len(s) - 1, s):
                memo[start] = 0
            else:
                for end in range(start, len(s) - 1):
                    if self.isPalindrome(start, end, s): #include end
                        memo[start] = min(memo[start], 1 + memo[end + 1])
        return memo[0]

    def isPalindrome(self, start, end, s):
        i = start
        j = end
        while i < j:
            if s[i] == s[j]:
                i += 1
                j -= 1
            else:
                return False
        return True

if __name__ == '__main__':
    mySolution = Solution()
    print(mySolution.minCut('abbccbba'))                    
                    
                    
                
